File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3129_find-all-possible-stable-binary-arrays-i/solutions/cpp/solution.cpp
#include <vector>
#include <iostream>

using namespace std;

class Solution {
public:
    int numberOfStableArrays(int zero, int one, int limit) {
        const int MOD = 1000000007;
        vector<vector<vector<int>>> dp(zero + 1, vector<vector<int>>(one + 1, vector<int>(limit + 1, 0)));
        
        // Base cases
        if (zero > 0) dp[1][0][1] = 1; // Array with a single 0
        if (one > 0) dp[0][1][1] = 1;  // Array with a single 1
        
        for (int z = 0; z <= zero; ++z) {
            for (int o = 0; o <= one; ++o) {
                for (int l = 1; l <= limit; ++l) {
                    if (z == 0 && o == 0) continue; // Skip the zero length array case
                    if (z + o == 1) continue;       // Skip already initialized base cases
                    
                    // If we can place another 0
                    if (z > 0) {
                        dp[z][o][l] = (dp[z][o][l] + (l > 1 ? dp[z - 1][o][l - 1] : dp[z - 1][o][min(o, limit)])) % MOD;
                    }
                    
                    // If we can place another 1
                    if (o > 0) {
                        dp[z][o][l] = (dp[z][o][l] + (l > 1 ? dp[z][o - 1][l - 1] : dp[z][o - 1][min(z, limit)])) % MOD;
                    }
                }
            }
        }
        
        int result = 0;
        for (int l = 1; l <= limit; ++l) {
            result = (result + dp[zero][one][l]) % MOD;
        }
        
        return result;
    }
};

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3129_find-all-possible-stable-binary-arrays-i/solutions/java/solution.java
import java.util.Arrays;

class Solution {
    private static final int MOD = 1000000007;

    public int numberOfStableArrays(int zero, int one, int limit) {
        // Using dynamic programming where dp[i][j] represents the count of stable arrays with
        // exactly i zeroes and j ones that end in the current state (0 or 1).
        // Specifically, dp0[i][j] - ends in 0, dp1[i][j] - ends in 1
        int[][] dp0 = new int[zero + 1][one + 1];
        int[][] dp1 = new int[zero + 1][one + 1];

        // Base cases
        if (zero > 0) dp0[1][0] = 1; // Single element array ending with 0
        if (one > 0) dp1[0][1] = 1; // Single element array ending with 1

        // Fill the DP tables
        for (int z = 0; z <= zero; z++) {
            for (int o = 0; o <= one; o++) {
                if (z == 0 && o == 0) continue; // Skip the empty case

                if (z > 0) {
                    // Accumulate results for ending with 0, considering a maximum of `limit` previous 1's
                    for (int k = 1; k <= limit && k <= o; k++) {
                        dp0[z][o] = (dp0[z][o] + dp1[z][o - k]) % MOD;
                    }
                }
                
                if (o > 0) {
                    // Accumulate results for ending with 1, considering a maximum of `limit` previous 0's
                    for (int k = 1; k <= limit && k <= z; k++) {
                        dp1[z][o] = (dp1[z][o] + dp0[z - k][o]) % MOD;
                    }
                }
            }
        }

        // Final result is the sum of configurations that end with either 0 or 1 with exact count of zero's and one's
        return (dp0[zero][one] + dp1[zero][one]) % MOD;
    }
}
